Permissions Layer for Versioned store

mnaamani opened the pull request:
WIP:

Implements https://github.com/Joystream/joystream/issues/92

- [x] Basic Implementation
- [x] Reconsider how initial class permissions are set when new class is created?
 - [x] Update versioned store - This module "needs" read-only access to `Entity.class_id`, `ClassPropertyValue.value`, `ClassPeropertyValue.in_class_index` but they are currently private fields (add some helper methods instead of direct access) - @siman ?
- [x] Tests

mnaamani committed:
adding types

mnaamani committed:
set_class_entity_permissions

mnaamani committed:
use updated versioned store

mnaamani committed:
rustfmt

mnaamani committed:
remaining methods for updating concrete permissions and setting admins

mnaamani committed:
adding some proxy calls to versioned store

mnaamani committed:
make entity ownership an Option

mnaamani committed:
updating entities

mnaamani committed:
check Internal reference constraints when updating entity properties

mnaamani committed:
implement traits for () and (X, Y) and some code cleanup

mnaamani committed:
fix match arm

mnaamani committed:
class_permissions provided when creating a class

mnaamani committed:
prepare mocks and tests

mnaamani committed:
add mock membership checker and can create classes checker

mnaamani committed:
tests

mnaamani committed:
refactor errors

mnaamani committed:
entity ownership transfer

mnaamani committed:
tests

mnaamani committed:
add From impl for constructing sets from vectors

mnaamani committed:
tests

mnaamani committed:
factoring out some code into modules

mnaamani committed:
refactoring into modules

mnaamani committed:
cleanup

mnaamani committed:
clippy suggestions

siman commented on src/permissions.rs:
```diff
@@ -0,0 +1,175 @@
+{
+    // concrete permissions
+    /// Permissions that are applied to entities of this class, defines who in addition to
+    /// root origin can update and delete entities of this class.
```
We cannot delete entities yet. (Will we be able to do this in the future? Not decided yet)
```suggestion
    /// root origin can update entities of this class.
```

siman commented on src/permissions.rs:
```diff
@@ -0,0 +1,175 @@
+    GroupId: Ord,
+{
+    pub update: EntityPrincipalSet<AccountId, GroupId>,
+    pub delete: EntityPrincipalSet<AccountId, GroupId>,
```
Versioned store doesn't have a functionality to delete something on Entity.
I commented it out as @bedeho asked
```suggestion
```

mnaamani replied:
done

siman commented on src/tests.rs:
```diff
@@ -0,0 +1,510 @@
+        // remove special permissions for entity owners
+        entity_permissions: EntityPermissions {
+            update: EntityPrincipalSet::new(),
+            delete: EntityPrincipalSet::new(),
```
Versioned store doesn't have a functionality to delete something on Entity.
I commented it out as @bedeho asked
```suggestion
```

mnaamani replied:
done

siman commented on src/tests.rs:
```diff
@@ -0,0 +1,510 @@
+            "NotInEntityPermissionsUpdateSet"
+        );
+
+        // default permissions give entity owner permission to update and delete
```
Versioned store doesn't have a functionality to delete something on Entity.
```suggestion
        // default permissions give entity owner permission to update
```

siman commented on src/tests.rs:
```diff
@@ -0,0 +1,510 @@
+        let class_permissions = Permissions::class_permissions_by_class_id(class_id);
+
+        assert!(class_permissions.entity_permissions.update.is_empty());
+        assert!(class_permissions.entity_permissions.delete.is_empty());
```
```suggestion
```

siman commented on src/tests.rs:
```diff
@@ -0,0 +1,510 @@
+            update: EntityPrincipalSet::from(vec![EntityPrincipal::Base(BasePrincipal::Account(
+                1,
+            ))]),
+            delete: EntityPrincipalSet::from(vec![EntityPrincipal::Base(BasePrincipal::Account(
```
Delete `delete` entry

siman commented on src/tests.rs:
```diff
@@ -0,0 +1,510 @@
+            update: EntityPrincipalSet::from(vec![EntityPrincipal::Base(BasePrincipal::Account(
+                4,
+            ))]),
+            delete: EntityPrincipalSet::from(vec![EntityPrincipal::Base(BasePrincipal::Account(
```
Delete `delete` entry

siman commented on src/permissions.rs:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    pub fn can_add_schema(
```
I suggest improving naming to fully reflect for what method this check is used.
```suggestion
    pub fn can_add_class_schema(
```

siman commented on src/permissions.rs:
```diff
@@ -0,0 +1,175 @@
+                {
+                    Ok(())
+                } else {
+                    Err("NotInEntityPermissionsUpdateSet")
```
```suggestion
                    Err("BasePrincipalIsNotInEntityPermissionsUpdateSet")
```

mnaamani replied:
renaming to `CredentialIsNotInEntityPermissionsUpdateSet`

siman commented on src/lib.rs:
```diff
@@ -0,0 +1,586 @@
+                            // is entity owned by system
+                            ensure!(
+                                <EntityOwnerByEntityId<T>>::exists(entity_id),
+                                "NotEnityOwner"
```
```suggestion
                                "PrincipalIsNotEnityOwner"
```

mnaamani replied:
outdated

siman commented on src/lib.rs:
```diff
@@ -0,0 +1,586 @@
+                                {
+                                    Ok(DerivedPrincipal::EntityOwner)
+                                }
+                                _ => Err("NotEnityOwner"),
```
```suggestion
                                _ => Err("PrincipalIsNotEnityOwner"),
```

mnaamani replied:
outdated

siman commented on src/lib.rs:
```diff
@@ -0,0 +1,586 @@
+                    // is entity owned by system?
+                    ensure!(
+                        <EntityOwnerByEntityId<T>>::exists(entity_id),
+                        "NotEnityOwner"
```
```suggestion
                        "PrincipalIsNotEnityOwner"
```

mnaamani replied:
outaded

siman commented on src/lib.rs:
```diff
@@ -0,0 +1,586 @@
+                        {
+                            Ok(DerivedPrincipal::EntityOwner)
+                        }
+                        _ => Err("NotEnityOwner"),
```
```suggestion
                        _ => Err("PrincipalIsNotEnityOwner"),
```

mnaamani replied:
outdated

siman commented on src/principles.rs:
```diff
@@ -0,0 +1,117 @@
+#[derive(Encode, Decode, Eq, PartialEq, Ord, PartialOrd, Clone, Debug)]
+pub enum EntityPrincipal<AccountId, GroupId> {
+    Base(BasePrincipal<AccountId, GroupId>),
+    Owner,
```
Should be called `EntityOwner` as in `DerivedPrincipal`?
```suggestion
    EntityOwner,
```

mnaamani replied:
outdated

siman commented on src/permissions.rs:
```diff
@@ -0,0 +1,175 @@
+    PropertyIndex: Ord,
+{
+    // concrete permissions
+    /// Permissions that are applied to entities of this class, defines who in addition to
```
```suggestion
    /// Permissions that are applied to entities of this class, define who in addition to
```

siman commented:
Seems like there is a typo in a file name: `principles.rs` -> `principals.rs`

siman commented on src/constraint.rs:
```diff
@@ -0,0 +1,28 @@
+    /// Any entity may reference the class.
+    NoConstraint,
+
+    /// Only a set of entities of type ClassId and from the specified property index can reference the class.
```
This comment seems a bit wrong because here you specify what exact properties (property indexes) on what exact class can refer to this class. There is nothing about entity in this field.

mnaamani replied:
fixed in 16c068976fd4a7a692c2a35661fba2298267ee41

siman commented on src/constraint.rs:
```diff
@@ -0,0 +1,28 @@
+use codec::{Decode, Encode};
+use rstd::collections::btree_set::BTreeSet;
+
+/// Pointer to a specific property of entities of a specfic class.
```
There is nothing about entity in this structure. You only refer to class and property index.
```suggestion
/// Pointer to a specific property of a specific class.
```

siman replied:
I mean that property is declared in class, whereas property value is a part of an entity.
So, in this case, it's just a property of a class.

siman commented on src/lib.rs:
```diff
@@ -0,0 +1,586 @@
+    ) -> Result<ClassPermissionsType<T>, &'static str> {
+        ensure!(
+            <ClassPermissionsByClassId<T>>::exists(class_id),
+            "ClassIdDoesNotExist"
```
```suggestion
            "ClassPermissionsNotFoundByClassId"
```

mnaamani replied:
fixed in 8c279444ab63e0a71f9ce742492c834b81c739cf

siman commented on src/principles.rs:
```diff
@@ -0,0 +1,117 @@
+#[derive(Encode, Decode, Eq, PartialEq, Ord, PartialOrd, Clone, Debug)]
+pub enum BasePrincipal<AccountId, GroupId> {
+    Account(AccountId),
+    GroupMember(GroupId),
```
```suggestion
    Group(GroupId),
    // or even `ExternalGroup` to emphasize that this group is defined out of the scope of this module:
    ExternalGroup(GroupId),
```

mnaamani replied:
outdated

mnaamani committed:
simplify concept of principals, update test

mnaamani committed:
owner -> maintainer

mnaamani committed:
remove set_entity_maintainer() method

bedeho commented:
I have not reviwed, let me know if I should, can make sense before you start making new major changes that we planned.

However I wanted to make a small ask, could we change from "participant" to "credential" ? You the reason is that the mapping of the id corresponds to such disperate actor(s)

- the lead
- one or more curators
- one or more members/channel owners
- any member

so really, what is being furnished when trying to write to the store is not so much your ID as a participant, but a sort of credential for what it is you want to do, and the WG holds the underlying information about what account actually holds what credential at any given time. That naming makes more sense to me.

Let me know what you think.

bedeho committed:
Missing inclusion of Substrate standard lib

Normal build hides this, since normal Rust standard library is added implicitly.

mnaamani committed:
update substrate dependency

mnaamani committed:
fixes after substrate update

mnaamani committed:
depend on versioned store from joystream repo

mnaamani committed:
update versioned store dependency

mnaamani commented:
I think you might be right about the naming of the id. Here is some [background about security terminology](https://docs.oracle.com/javase/7/docs/technotes/guides/security/jgss/tutorials/glossary.html) related to prinicpals,subjects,authentication and credentials.

referencing from link above: "Credentials might also contain data that enables the subject to perform certain activities", it seems to translate well to what these IDs represent.

mnaamani committed:
refactor PrincipalId -> Credential

mnaamani committed:
refactor to pass raw_origin to helper methods, preparing batching operations

mnaamani committed:
batching: create_entity and update_property_values

mnaamani committed:
batching: add_schema_support_to_entity

mnaamani committed:
batching: add ParametrizedEntity

mnaamani committed:
batching tests

mnaamani committed:
support parametrized vector of entities

mnaamani committed:
do not type alias u16 - causes problem with polkadot-js/api

mnaamani committed:
Update src/permissions.rs

Co-Authored-By: Alex Siman <aleksandr.siman@gmail.com>

mnaamani committed:
Update src/constraint.rs

Co-Authored-By: Alex Siman <aleksandr.siman@gmail.com>

mnaamani committed:
Update src/permissions.rs

Co-Authored-By: Alex Siman <aleksandr.siman@gmail.com>

mnaamani committed:
Update src/permissions.rs

Co-Authored-By: Alex Siman <aleksandr.siman@gmail.com>

mnaamani committed:
use renamed can_add_class_schema method

mnaamani committed:
improved comments for ReferenceConstraint

mnaamani committed:
better Err strings

mnaamani commented:
Incorporated most of your suggestions @siman although since the review was a bit out of date some of the review comments were not relevant any more.
Ready for final review.

mnaamani commented:
Also just a final thought, since we are not going to invest more in features for this module beyond rome release there is no point making any additional improvements, so only bug fixes really.
